verifier/crl/crl.go

```diff
@@ -0,0 +1,195 @@
+	logger.Infof("Retrieving crl bundle from file cache with key %q ...", url)
+
+	// get content from file cache
+	f, err := os.Open(filepath.Join(c.root, c.fileName(url)))
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			logger.Infof("CRL file cache miss. Key %q does not exist", url)
+			return nil, corecrl.ErrCacheMiss
+		}
+		return nil, fmt.Errorf("failed to get crl bundle from file cache with key %q: %w", url, err)
+	}
+	defer f.Close()
+
+	// decode content to crl Bundle
+	var content fileCacheContent
+	err = json.NewDecoder(f).Decode(&content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode file retrieved from file cache: %w", err)
+	}
+	var bundle corecrl.Bundle
+	bundle.BaseCRL, err = x509.ParseRevocationList(content.BaseCRL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse base CRL of file retrieved from file cache: %w", err)
+	}
+	if content.DeltaCRL != nil {
+		bundle.DeltaCRL, err = x509.ParseRevocationList(content.DeltaCRL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse delta CRL of file retrieved from file cache: %w", err)
+		}
+	}
+
+	// check expiry
+	if err := checkExpiry(ctx, bundle.BaseCRL.NextUpdate); err != nil {
+		return nil, err
+	}
+	if bundle.DeltaCRL != nil {
+		if err := checkExpiry(ctx, bundle.DeltaCRL.NextUpdate); err != nil {
+			return nil, err
+		}
+	}
+
+	return &bundle, nil
+}
+
+// Set stores the CRL bundle in c with url as key.
+func (c *FileCache) Set(ctx context.Context, url string, bundle *corecrl.Bundle) (setErr error) {
+	logger := log.GetLogger(ctx)
+	logger.Infof("Storing crl bundle to file cache with key %q ...", url)
+
+	// sanity check
+	if bundle == nil {
+		return errors.New("failed to store crl bundle in file cache: bundle cannot be nil")
+	}
+	if bundle.BaseCRL == nil {
+		return errors.New("failed to store crl bundle in file cache: bundle BaseCRL cannot be nil")
+	}
+
+	// actual content to be saved in the cache
+	content := fileCacheContent{
+		BaseCRL: bundle.BaseCRL.Raw,
+	}
+	if bundle.DeltaCRL != nil {
+		content.DeltaCRL = bundle.DeltaCRL.Raw
+	}
+
+	// save content to temp file
+	tempFile, err := os.CreateTemp("", tempFileName)
+	if err != nil {
+		return fmt.Errorf("failed to store crl bundle in file cache: failed to create temp file: %w", err)
+	}
+	defer func() {
+		// remove the temp file in case of error
+		if setErr != nil {
+			defer os.Remove(tempFile.Name())
+		}
+	}()
+
+	if err := json.NewEncoder(tempFile).Encode(content); err != nil {
+		return fmt.Errorf("failed to store crl bundle in file cache: failed to encode content: %w", err)
+	}
+	if err := tempFile.Close(); err != nil {
+		return fmt.Errorf("failed to store crl bundle in file cache: failed to close temp file: %w", err)
+	}
+
+	// rename is atomic on UNIX-like platforms
+	if err := os.Rename(tempFile.Name(), filepath.Join(c.root, c.fileName(url))); err != nil {
+		return fmt.Errorf("failed to store crl bundle in file cache: %w", err)
+	}
```

Comment:
You may want to refactor this to simplify the code flow. We can chat later offline.

Reply:
Updated after discussion.